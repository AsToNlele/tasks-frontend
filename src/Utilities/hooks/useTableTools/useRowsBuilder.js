const columnProp = (column) =>
  column.key ||
  column.original?.toLowerCase() ||
  column.sortByProp?.toLowerCase();

const buildRow = (item, columns, index) =>
  columns.map((column) => ({
    title: column.renderFunc
      ? column.renderFunc(undefined, undefined, item, index)
      : item[columnProp(column)],
  }));

const useRowsBuilder = (items, columns, options = {}) => {
  const filteredItems = options?.filter ? options.filter(items) : items;

  const sortedItems = options?.sorter
    ? options.sorter(filteredItems)
    : filteredItems;

  const paginatedItems = options?.paginator
    ? options?.paginator(filteredItems)
    : sortedItems;

  const rows = paginatedItems.map((item, index) =>
    buildRow(item, columns, index)
  );

  const pagination = options?.pagination
    ? {
        ...options.pagination,
        itemCount: filteredItems.length,
      }
    : undefined;

  return {
    tableProps: { rows },
    toolbarProps: {
      pagination,
    },
  };
};

export default useRowsBuilder;
