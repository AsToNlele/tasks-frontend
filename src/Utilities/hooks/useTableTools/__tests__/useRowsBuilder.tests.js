import { renderHook } from '@testing-library/react-hooks';
import items from './__fixtures__/items.fixtures';
import columns from './__fixtures__/columns.fixtures';
import useRowsBuilder from '../useRowsBuilder';

describe('useRowsBuilder', () => {
  it('returns a rows configuration', () => {
    const { result } = renderHook(() =>
      useRowsBuilder(items, columns, {
        rowTransformer: [undefined],
      })
    );
    expect(result).toMatchSnapshot();
  });
});
