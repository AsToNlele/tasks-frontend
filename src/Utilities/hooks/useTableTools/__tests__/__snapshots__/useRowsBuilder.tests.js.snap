// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRowsBuilder returns a rows configuration 1`] = `
Object {
  "all": Array [
    Object {
      "tableProps": Object {
        "rows": Array [
          Array [
            Object {
              "title": "taskA",
            },
            Object {
              "title": 10,
            },
            Object {
              "title": "Running",
            },
          ],
          Array [
            Object {
              "title": "taskB",
            },
            Object {
              "title": 5,
            },
            Object {
              "title": "20 Apr 2022, 11:10 UTC",
            },
          ],
          Array [
            Object {
              "title": "taskB",
            },
            Object {
              "title": 3,
            },
            Object {
              "title": "25 Apr 2022, 11:10 UTC",
            },
          ],
        ],
      },
      "toolbarProps": Object {
        "pagination": undefined,
      },
    },
  ],
  "current": Object {
    "tableProps": Object {
      "rows": Array [
        Array [
          Object {
            "title": "taskA",
          },
          Object {
            "title": 10,
          },
          Object {
            "title": "Running",
          },
        ],
        Array [
          Object {
            "title": "taskB",
          },
          Object {
            "title": 5,
          },
          Object {
            "title": "20 Apr 2022, 11:10 UTC",
          },
        ],
        Array [
          Object {
            "title": "taskB",
          },
          Object {
            "title": 3,
          },
          Object {
            "title": "25 Apr 2022, 11:10 UTC",
          },
        ],
      ],
    },
    "toolbarProps": Object {
      "pagination": undefined,
    },
  },
  "error": undefined,
}
`;
